### ISSUE TITLE
1: Getting Started (Practical Deep Learning for Coders)

### ISSUE THREAD

**Issue by JeremyLoh** [JeremyLoh/ml#5]
https://course.fast.ai/Lessons/lesson1.html

## Own Kaggle notebook for Chapter 1 of book
- https://www.kaggle.com/jeremyloh42/chapter-1-deep-learning-for-coders/

# Reference Book
- [Deep Learning for Coders with Fastai and PyTorch: AI Applications Without a PhD](https://www.amazon.com/Deep-Learning-Coders-fastai-PyTorch/dp/1492045527) https://course.fast.ai/Resources/book.html
- Other book - [Natural Language Processing with Transformers Book](https://transformersbook.com/) https://transformersbook.com/

## Tasks
- [x] Machine Learning: Train vs. Validation vs. Test Sets https://www.youtube.com/watch?v=PIRQY6xmNZY
- [x] Read about learning rate: Using the Learning Rate Finder (Beginner) - https://walkwithfastai.com/lr_finder
- [x] **What is overfitting?** https://aws.amazon.com/what-is/overfitting/
- [x] What and how does show_batch actually do? https://forums.fast.ai/t/what-and-how-does-show-batch-actually-do/90014
    - This is a high level answer - but essentially show_batch goes through all the transforms you’ve provided explicitly or via the DataBlock, and then keeps reversing through them until it finds a Transform that has a decode method that returns an object that can show itself. Typically the show method will return a matplotlib object that can be displayed. Sorry it’s not more detailed, but you can dig in deeper here https://docs.fast.ai/tutorial.transformers.html
- [x] Finish Chapter 1 of book (Deep Learning for Coders with Fastai and PyTorch: AI Applications Without a PhD)
    -  **Current Page - 138** 
    - Page 111 
![Image](https://user-images.githubusercontent.com/38811217/261834490-b1f9a8f0-4827-4bb9-818b-dccc6d408223.png)
    - Page 69 / 1028 - The error rate is the proportion of images that were incorrectly identified
    - Page 77 - Neural network is a flexible function that can solve any problem to any level of accuracy, in theory, based on its weights. (proof is the universal approximation theorem). 
        - How do we adjust the weights automatically? Stochastic Gradient Descent (SGD) 
    - Page 80 - The results of the model are called predictions. The measure of performance is called the loss (loss depends not only on the predictions, but also on the correct labels (targets / dependent variable) (e.g. "cat" or "dog").
        - We need not just input data, but also labels for that data 
    - Page 81 - Feedback loops can be created as the model interacts with its environment.
        - e.g. Predictive policing model is created based on where arrests have been made in the past. (reflecting biases in existing policing processes)
        - Law enforcement officers might use that model to decide where to focus their policing activity, resulting in increased arrests in those areas
        - Data on these additional arrests would be fed back into retrain future versions of the model
        - This is an example of a **positive feedback loop**. The more the model is used, the more biased the data becomes, making the model even more biased 
    - Computer vision datasets are normally structured in a way that the label for an image is part of the filename or path - most commonly the parent folder name
    - Page 86 - Transform contains code that is applied automatically during training. item_tfms are applied to each item, batch_tfms are applied to a batch of items at a time using the GPU
        - Resize image size - 224 pixels is for historical reasons. Increasing the size, you will often get a model with better results (it can focus on more details), but at the price of speed and memory consumption. Opposite effect if we decrease the image size
        - Classification model is one that attempts to predict a class / category (predict from a number of discrete possibilities)
        - Regression model is one that attempts to predict one or more numeric quantities, e.g. temperature or location  
    - Page 87 - valid_pct=0.2 means that 20% of the data is held for validation set, used to measure the accuracy of the model. 80% of the data is used as training set. fastai shows the model's accuracy only using the validation set, not the training set. We only care about how the model works on **previously unseen images**
        - The longer you train for, the model might be memorize the training set data, instead of finding generalizable underlying patterns in the data. (The model is **overfitting**)
        - Overfitting is the single most important and challenging issue when training. We should only use methods to avoid overfitting only after you have confirmed that overfitting is occurring (the validation accuracy is getting worse during training). If you train for too long, with not enough data, you will see the accuracy of your model start to get worse (overfitting)
    - Page 91 CNN (Convolutional Neural Network) is the current state of the art approach to creating computer vision models.
        - We need to specify what **architecture** to use (what kind of model to create), what data we want to train on (dls, data loader), and what **metric** to use
        - e.g. architecture, ResNet, resnet34 -> 34 refers to the number of layers in this variant of the architecture. e.g. other options are 18, 50, 101, 152. The more layers, the longer to train, and more prone to overfitting (can't train them for as many epochs before the accuracy on the validation set gets worse). When using more data, they can be more accurate.  
        - Metric is a function that measures the quality of the model's predictions using the validation set. It will be printed at the end of each epoch. e.g. **error_rate** tells the percentage of images in the validation set that are classified incorrectly. **accuracy** is just 1.0 - error_rate. 
        - **There is a difference between a metric and loss**. The purpose of **loss** is to define a "measure of performance" that the training system can use to update weights automatically. (a good choice for loss is a choice that is easy for stochastic gradient descent to use). A **metric** is defined for human consumption, one that is easy for us to understand, and as closely as possible to what you want the model to do.
        - When using a Pretrained model - cnn_learner will remove the last layer, since it is always specifically customized to the original training task, and replace it with one or more new layers with randomized weights, of an appropriate size for the dataset you are working with (this last part of the model is known as the **head**)
        - Using a pretrained model for a task different from what it was originally trained for is known as **transfer learning**
        - **Fine Tuning** is a transfer learning technique that updates the parameters of a pretrained model by training for additional epochs using a different task from that used for pretraining
        - There is a "fit" method, which fit a model (e.g. look at images in the training set multiple times, each time updating the parameters to make the predictions closer to the target labels). But this would throw away all capabilities that a pretrained model has.
        - ![Image](https://user-images.githubusercontent.com/38811217/261833463-1d99b48e-0a20-42ac-9267-ef4c119f1868.png)
      - Segmentation is being able to recognize the context of every individual pixel in an image 
      - Page 126: training set (data model see in training), validation set (only use for evaluation, let us test the model, in training phase), test set (used after the model is fully trained to access model performance). This is needed to reduce change of overfitting. (hyperparameters - exploring various versions of a model through modelling choices). We might overfit the validation data through human trial and error, when trying to explore different models. **We hold back the test data, it cannot be used to improve the model. It can only be used to evaluate the model at the end of all training**
      - page 129 - for time series data, we want to choose a continuous section with the latest dates in your validation set, and older dates for training set.
      - ![image](https://github.com/JeremyLoh/ml/assets/38811217/59bd4ccf-6a3c-415d-8217-8202b4c20e3c)

In this lesson you’re going to hit the ground running – in the first five minutes you’ll see a complete end to end example of training and using a model that’s so advanced it was considered at the cutting edge of research capabilities in 2015.

- [x] 1:07:00 / 1:22:55 Lesson 1: Practical Deep Learning for Coders 2022 https://www.youtube.com/watch?v=8SF_h3xF3cE

# 2:00 PixSpy - https://pixspy.com/
- Image inspection multi tool

# 11:45 https://ethics.fast.ai/

# 14:30 Meta Learning: How To Learn Deep Learning And Thrive In The Digital World https://rosmulski.gumroad.com/l/learn_machine_learning

# 27:40 Using Image classifier to classify sounds (image of their waveforms)

# 51:00 https://timm.fast.ai/

# 55:00 Computer vision is not just for image recognition. e.g. Segmentation => classify what is present in a picture, e.g. cars, people. Uses UNET learner

# 58:20 Tabular analysis (taking spreadsheets / database tables, and trying to predict columns of those) - e.g. income prediction
- fast ai has function `untar_data` is to download some data and decompresses it for you, has common urls for dataset that we can use. e.g. URLs.ADULT_SAMPLE

![Image](https://user-images.githubusercontent.com/38811217/261787029-e4340fb2-9bc4-4845-bd6c-e23a51c68602.png)

  - `from fastai.tabular.all import untar_data`
  - cat_names -> tell which columns are categorical (discrete values)
  - cont_names -> tell which columns are continuous values (any real number)

dls.show_batch() to see the data generated (dls is dataloaders)

For tabular models, we do 
```python
learn = tabular_learner(dls, metrics=accuracy)
learn.fit_one_cycle(2) # we use this instead of fine tune as there isn't going to be a pre-trained model to do what you want
# every table will be different
```

# 1:01:10 Collaborative Filtering - e.g. Recommendation system
System where we take a data set that says which users liked which products / used which products, => then use that to guess what other products those users might like based on finding similar users and what those similar users like
- For the learning, the range we are predicting is some range of real numbers. e.g. the actual range is 1 to 5, but we will go a bit lower for the min range, and a little higher for the max range. Reasons will be explained later
```python
learn = collab_learner(dls, y_range=(0.5, 5.5))
learn.fine_tune(10) # actually we should use fit_one_cycle as there isn't really a pre-trained model
```

![Image](https://user-images.githubusercontent.com/38811217/261787451-ce67f9f3-1aa9-4d40-a590-6e5373a662cc.png)
- valid_loss is the mean squared error. on average how far off are we, for the validation set

For fast.ai library, we can do `learn.show_results()` to get some sensible results
![Image](https://user-images.githubusercontent.com/38811217/261787537-d8efa9d5-4fdb-43dd-9100-79104c713e83.png)

If you have questions about a fastai method, you should use the function `doc`, passing in the method name. e.g. `doc(learn.predict`

# 1:05:40 Jupyter notebooks extension, RISE

Transfer learning
